Addressing issue 89: incorporating SUP on App Registry as OCI Registry (#123)
margo/specification: system-design/concepts/applications/local-registries.md

# Local Registries

> The [Component Registry](application-registry.md) as well as the [Container Registry](application-registry.md) considered by Margo are the registries that need to be accessed by devices to download components (i.e., Helm Charts or Compose Archives) or containers. This section describes how to configure those as local registries to avoid reliance on public Internet-accessible registries.

Reviewer: ```suggestion
The [Component Registry](application-registry.md) as well as the [Container Image Registry](application-registry.md) considered by Margo are the registries that need to be accessed by devices to download components (i.e., Helm Charts or Compose Archives) or container images.
This section describes how to configure those as local registries to avoid reliance on public Internet-accessible registries.
```

Author: done

The reasons for not using public registries are for example: (1) publicly hosted container images or Helm charts could become unavailable at some point, as the owner decides to take the container images or Helm charts off the public registry, (2) Internet connectivity may not be available to the device and hence public registries are not reachable, or (3) end-users want to host their own registries so they can do security scans and package validation.

In terms of connectivity, we can thereby distinguish mainly between the following device categories:

1. **Fully connected device**, which means a device is deployed in the field (e.g., a factory shop floor) and has access to the Internet.
2. **Locally connected device**, i.e., the device has connectivity to a local network (e.g., factory- or enterprise-wide) and a local repository can be made reachable.
3. **Air-gapped device**, i.e., the device generally is not connected and must be configured by accessing it directly (via direct network link with an Ethernet cable, or similar) for example through a technician’s laptop.

> Local Component Registries or Container Registries can be used for all device categories. In case of **fully connected devices**, although the device could reach the Internet, a local registry can still be useful (e.g., as a cache for remote registries to save on bandwidth or to have container images or Helm Charts reliably available). In case of **locally connected devices**, a local registry is required to enable the Workload Fleet Management Client (WFMC) to install a Margo application on the device, as the device/WFMC does not have Internet access. Thereby, the local registry can be set up as a *pull-through cache* where data (e.g., container images) is cached locally when they are first retrieved from a remote source and subsequent requests for the same data are served from the local cache rather than fetching it again from the remote source. In case of **air-gapped devices**, a local registry has to be accessible on the technician's laptop (or other directly connected device), which performs the application installation process.

Reviewer: There is no need to provide an OCI registry on air-gapped devices. All artifacts provided as OCI artifacts (container images, Margo application,...) can be exported to an archive that can be easily transferred to the connected device (e.g. with a USB-stick) which then imports them. But I don't know how to put that information in the text, therefore is not part of the below "suggestion".

```suggestion
Local Component Registries or Container Image Registries can be used for all device categories.
In case of **fully connected devices**, although the device could reach the Internet, a local registry can still be useful (e.g., as a cache for remote registries to save on bandwidth or to have container images or Helm Charts reliably available).
In case of **locally connected devices**, a local registry is required to enable the Workload Fleet Management Client (WFMC) to install a Margo application on the device, as the device/WFMC does not have Internet access.
Thereby, the local registry can be set up as a *pull-through cache* where data (e.g., container images) is cached locally when they are first retrieved from a remote source and subsequent requests for the same data are served from the local cache rather than fetching it again from the remote source.
In case of **air-gapped devices**, a local registry has to be accessible on the technician's laptop (or other directly connected device), which performs the application installation process.
```

Author: done

To setup local registries, different configuration options exist:

> ## Option - Container Registry Mirror on Kubernetes Level

Reviewer: ```suggestion
## Option - Container Image Registry Mirror on Kubernetes Level
```

Author: done

Kubernetes supports the configuration of registry mirrors. How this is configured depends on the distribution and the underlying container runtime. Distributions that utilize **containerd** as runtime (e.g., k3s or microk8s) allow the definition of mirrors in a configuration file. For example, in k3s the file `/etc/rancher/k3s/registries.yaml` can be used to set up a mirror for each device's Kubernetes environment:

```yaml
mirrors:
  "docker.io":
    endpoint:
      - "http://<local-registry-ip>:5000"
configs:
  "docker.io":
    auth:
      username: "<username>"
      password: "<password>"
```

> ## Option - Container Registry as Pull-through Cache on Docker Level

Reviewer: ```suggestion
## Option - Container Image Registry as Pull-through Cache on Docker Level
```

Author: done

> To configure a pull-through cache in Docker for the Container Registry, a Docker Registry can be setup that acts as caching proxy for a remote Docker Registry. Such a Docker Registry can be defined using the following `config.yml`:

Reviewer: ```suggestion
To configure a pull-through cache in Docker for the Container Image Registry, a Docker Registry can be setup that acts as caching proxy for a remote Docker Registry. Such a Docker Registry can be defined using the following `config.yml`:
```

Author: done
